src/feature-extraction/features.jl

```diff
@@ -1,494 +1,362 @@
-    FeatureValue(posf(rec[pastframe][vehicle_index].state).s)
-end
-function Base.get(::Feature_PosFt, rec::EntityQueueRecord{S,D,I}, roadway::R, vehicle_index::Int, pastframe::Int=0) where {S,D,I,R}
-    FeatureValue(posf(rec[pastframe][vehicle_index].state).t)
-end
-function Base.get(::Feature_PosFyaw, rec::EntityQueueRecord{S,D,I}, roadway::R, vehicle_index::Int, pastframe::Int=0) where {S,D,I,R}
-    FeatureValue(posf(rec[pastframe][vehicle_index].state).ϕ)
-end
-function Base.get(::Feature_Speed, rec::EntityQueueRecord{S,D,I}, roadway::R, vehicle_index::Int, pastframe::Int=0) where {S,D,I,R}
-    FeatureValue(vel(rec[pastframe][vehicle_index].state))
-end
-
-function Base.get(::Feature_VelFs, scene::Frame, roadway::R, vehicle_index::Int, pastframe::Int=0) where {R}
-    veh = scene[vehicle_index]
-    FeatureValue(velf(veh.state).s)
-end
-function Base.get(f::Feature_VelFs, rec::EntityQueueRecord{S,D,I}, roadway::R, vehicle_index::Int, pastframe::Int=0) where {S,D,I,R}
-    get(f, rec[pastframe], roadway, vehicle_index)
-end
-
-function Base.get(::Feature_VelFt, scene::Frame, roadway::R, vehicle_index::Int, pastframe::Int=0) where {R}
-    veh = scene[vehicle_index]
-    FeatureValue(velf(veh.state).t)
-end
-function Base.get(f::Feature_VelFt, rec::EntityQueueRecord{S,D,I}, roadway::R, vehicle_index::Int, pastframe::Int=0) where {S,D,I,R}
-    get(f, rec[pastframe], roadway, vehicle_index)
-end
-
-generate_feature_functions("TurnRateG", :turnrateG, Float64, "rad/s")
-function Base.get(::Feature_TurnRateG, rec::EntityQueueRecord{S,D,I}, roadway::R, vehicle_index::Int, pastframe::Int=0; frames_back::Int=1) where {S,D,I,R}
-
-    id = rec[pastframe][vehicle_index].id
-
-    retval = FeatureValue(0.0, FeatureState.INSUF_HIST)
-    pastframe2 = pastframe - frames_back
-    if pastframe_inbounds(rec, pastframe2)
-
-        veh_index_curr = vehicle_index
-        veh_index_prev = findfirst(id, rec[pastframe2])
-
-        if veh_index_prev != nothing
-            curr = posg(rec[pastframe][veh_index_curr].state).θ
-            past = posg(rec[pastframe2][veh_index_prev].state).θ
-            Δt = get_elapsed_time(rec, pastframe2, pastframe)
-            retval = FeatureValue(deltaangle(past, curr) / Δt)
-        end
+## define features traits
+
+# most general feature(roadway, scenes, actions, vehid)
+abstract type AbstractFeature end
+
+# feature(roadway, veh)
+struct EntityFeature <: AbstractFeature end
+# feature(roadway, scene, veh)
+struct FrameFeature <: AbstractFeature end
+# feature(roadway, scenes, veh)
+struct TemporalFeature <: AbstractFeature end
+# feature(actions, veh)
+# struct TemporalActionFeature <: AbstractFeature end 
+# feature(action, veh)
+# struct ActionFeature <: AbstractFeature end
+
+"""
+    featuretype(f::Function)
+function used for dispatching feature types depending on the method of the feature function 
+"""
+function featuretype(f::Function)
```

Comment:
No! :D just implement their function with one of the supported method. 
The trait is determined at **compile time**  so there should be no overhead.

Reply:
What I meant is a new `FeatureType`, such as `ActionFeature`?

Reply:
oh yes then, you would need to add a check for the new method.

Reply:
My main concern is that it's not easy right now to add actions to the interface (from the user side), so our callbacks effectively don't support actions

Reply:
the callbacks do support actions, the feature extraction does not for now. 

I think it would require extending the top level function `extract_features` to have a list of actions as inputs.

Reply:
Fair enough we can still use callbacks for collecting action related metrics, that should be good enough - didn't think about that